harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,440 @@
+
+        metrics_queue: queue.Queue = queue.Queue()
+        self._collector = _Collector(
+            metrics_queue=metrics_queue,
+            collection_interval=interval,
+        )
+        self._shipper = _Shipper(
+            session=self._session,
+            agent_id=self._agent_id,
+            trial_id=self._trial_id,
+            run_id=self._run_id,
+            metrics_queue=metrics_queue,
+        )
+
+        self._collector.start()
+        self._shipper.start()
+        self._on = True
+
+    def off(self) -> None:
+        """
+        Turns off profiling.
+
+        .. note::
+
+            This method is idempotent: if profiling is already off, this method is effectively a
+            no-op.
+        """
+        if not self._on:
+            return
+
+        logger.info("Stopping system metrics profiling.")
+        self.close()
+        self._on = False
+
+    def close(self) -> None:
+        """
+        Shuts down any threads that were created.
+        """
+        if self._collector:
+            self._collector.stop()
+        if self._shipper:
+            self._shipper.stop()
+
+
+class DummyProfilerContext(ProfilerContext):
+    """
+    An empty, drop-in replacement of ``ProfilerContext`` for cases when profiling cannot run.
+    """
+
+    def __init__(self) -> None:
+        pass
+
+    def on(self, interval: int = 1) -> None:
+        pass
+
+    def off(self) -> None:
+        pass
+
+    def close(self) -> None:
+        pass
+
+
+class _MetricGroupCollector(metaclass=abc.ABCMeta):
+    """
+    Abstract class that is subclassed by metric groups that implement their respective metrics
+    collection logic.
+    """
+
+    group: str
+
+    @abc.abstractmethod
+    def collect(self) -> typing.Dict[str, typing.Any]:
+        """
+        Sample and collect all metrics for this group and return them as a dictionary mapping
+        each metric name to its metric value.
+
+        Certain metrics may be associated with additional labels (i.e. GPU UUIDs) in which case
+        the returned dictionary should be nested with the label as keys. For example:
+
+        .. code:: python
+            {
+                "GPU-UUID-1": {
+                    "gpu_util": 0.12,
+                    "gpu_free_memory": 123.45,
+                },
+                "GPU-UUID-2": {
+                    "gpu_util": 0.23,
+                    "gpu_free_memory": 234.56,
+                }
+            }
+        """
+        pass
+
+
+class _Network(_MetricGroupCollector):
+    group = "network"
+
+    def __init__(self) -> None:
+        # Set initial values for throughput calculations.
+        self._start_ts = time.time()
+        self._start_vals = psutil.net_io_counters()
+
+        super().__init__()
+
+    def collect(self) -> typing.Dict[str, typing.Any]:
+        ts = time.time()
+        vals = psutil.net_io_counters()
+
+        sent_thru = (vals.bytes_sent - self._start_vals.bytes_sent) / (ts - self._start_ts)
```

Comment:
units here are bytes per second. This seems reasonable, but I wish it was easier to figure out without reading code.

Bytes per second since the profiler was turned on seems slightly less principled. Since the `core.Context` is public and the profiler is stored in a public variable in there, I think this means that the ProfilerContext is public API. Which I think itself means that we can expect users to `profiler.on` and `profiler.off`, themselves.

Is this a reasonable way to look at it?

Reply:
note to Wes: would I have been less confused with a word different than "start'?

Reply:
`interval_start_*` is currently my favorite. That these things are initially set during `__init__` and are being called `start` makes them look immutable to me.